File: C++/task2-3.cpp
#include <iostream>
#include <vector>
#include <tuple>

using namespace std;

int NODwithCoefs(int c, int m, int& x, int& y) 
{
    if (m == 0) 
    {
        x = 1;
        y = 0;
        return c;
    }

    int x1, y1;
    int nod = NODwithCoefs(m, c % m, x1, y1);

    x = y1;
    y = x1 - (c / m) * y1;

    return nod;
}

int ReversedNum(int c, int m) 
{
    int x, y;
    int gcd = NODwithCoefs(c, m, x, y);
    if (gcd != 1) 
    {
        return -1;
    } 
    else 
    {
        return (x % m + m) % m;
    }
}

int main() {
    int c, d, m;
    cout << "Введите c и m через пробел" << endl;
    cin >> c >> m;
    if (ReversedNum(c, m) != -1) 
    {
        d = ReversedNum(c, m);
        cout << "Обратное число " << c << "^(-1) mod " << m << " = " << d << endl;

        cout << "Остаток (r)\tX\t\tY\tЧастное (q)" << endl;
        vector<tuple<int, int, int, int>> table;
        int a = c, b = m;
        while (b != 0) 
        {
            int q = a / b;
            int r = a % b;
            int x, y;
            NODwithCoefs(a, b, x, y);
            table.push_back({r, x, y, q});
            a = b;
            b = r;
        }
        for (auto n : table) 
        {
            cout << get<0>(n) << "\t\t" << get<1>(n) << "\t\t" << get<2>(n) << "\t\t" << get<3>(n) << endl;
        }
    }
    else 
    {
        cout << "Для данных значений обратного числа d по модулю не существует." << endl;
    }
}

File: Java/task2-3.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    
    public static int NODwithCoefs(int c, int m, int[] xy) {
        if (m == 0) {
            xy[0] = 1;
            xy[1] = 0;
            return c;
        }
        
        int[] xy1 = new int[2];
        int nod = NODwithCoefs(m, c % m, xy1);
        
        xy[0] = xy1[1];
        xy[1] = xy1[0] - (c / m) * xy1[1];
        
        return nod;
    }
    
    public static int ReversedNum(int c, int m) {
        int[] xy = new int[2];
        int gcd = NODwithCoefs(c, m, xy);
        if (gcd != 1) {
            return -1;
        } else {
            return (xy[0] % m + m) % m;
        }
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Введите c и m через пробел");
        int c = scanner.nextInt();
        int m = scanner.nextInt();
        
        if (ReversedNum(c, m) != -1) {
            int d = ReversedNum(c, m);
            System.out.println("Обратное число " + c + "^(-1) mod " + m + " = " + d);
            
            System.out.println("Остаток (r)\tX\t\tY\tЧастное (q)");
            List<int[]> table = new ArrayList<>();
            int a = c, b = m;
            while (b != 0) {
                int q = a / b;
                int r = a % b;
                int[] xy = new int[2];
                NODwithCoefs(a, b, xy);
                table.add(new int[]{r, xy[0], xy[1], q});
                a = b;
                b = r;
            }
            for (int[] n : table) {
                System.out.println(n[0] + "\t\t" + n[1] + "\t\t" + n[2] + "\t\t" + n[3]);
            }
        } else {
            System.out.println("Для данных значений обратного числа d по модулю не существует.");
        }
    }
}
